nexus/src/app/background/instance_watcher.rs

```diff
@@ -0,0 +1,635 @@
+                        error: None,
+                        instance_updated: None,
+                    };
+                }
+                Err(e) => {
+                    slog::warn!(
+                        opctx.log,
+                        "error checking up on instance";
+                        "error" => ?e,
+                        "status" => ?e.status(),
+                    );
+                    return CheckResult {
+                        target,
+                        check_failure: None,
+                        error: Some(CheckError::ClientError),
+                        instance_updated: None,
+                    };
+                }
+            };
+
+            slog::debug!(opctx.log, "updating instance state: {state:?}");
+            let update_result = crate::app::instance::notify_instance_updated(
+                &datastore,
+                &resolver,
+                &opctx,
+                &opctx,
+                &opctx.log,
+                &instance_id,
+                &state.into(),
+            )
+            .await
+            .map_err(|_| CheckError::UpdateFailed)
+            .and_then(|updated| updated.ok_or(CheckError::InstanceNotFound));
+            match update_result {
+                Ok(updated) => {
+                    slog::debug!(opctx.log, "update successful"; "instance_updated" => updated.instance_updated, "vmm_updated" => updated.vmm_updated);
+                    CheckResult {
+                        target,
+                        instance_updated: Some(updated),
+                        check_failure: None,
+                        error: None,
+                    }
+                }
+                Err(e) => {
+                    slog::warn!(opctx.log, "error updating instance"; "error" => ?e);
+                    CheckResult {
+                        target,
+                        instance_updated: None,
+                        check_failure: None,
+                        error: Some(e),
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[derive(
+    Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, oximeter::Target,
+)]
+struct InstanceTarget {
+    /// The instance's ID.
+    instance_id: Uuid,
+    /// The sled-agent's ID.
+    sled_agent_id: Uuid,
+    /// The sled agent's IP address.
+    sled_agent_ip: IpAddr,
+    /// The sled agent's port.
+    sled_agent_port: u16,
+}
+
+struct CheckResult {
+    target: InstanceTarget,
+    /// `Some` if the instance's state was up
+    instance_updated: Option<crate::app::instance::InstanceUpdated>,
+
+    /// `Some` if the instance check indicated that the instance is in a bad state.
+    ///
+    /// This is a result that indicates that something is *wrong* with either the
+    /// sled on which the instance is running, the sled-agent on that sled, or the
+    /// instance itself. This is distinct from a [`CheckError`], which indicates
```

Comment:
Does this enum really reflect the instance state itself? From the variants below, it appears to be only about errors the sled-agent may return to us, not "something wrong with the instance itself."